src/Rbac.php

```diff
@@ -82,29 +84,109 @@ public function hasRole($role) : bool
             );
         }
 
-        if (is_string($role)) {
-            return isset($this->roles[$role]);
+        return $this->roleSearchIncludingChildren($this->roles, $role);
+    }
+
+    /**
+     * @param $obj|Array
+     * @param $needle|String
+     * @return bool
+     */
+    private function roleSearchIncludingChildren($obj, $needle) : bool
+    {
+        $rv = 0;
+
+        if (is_array($obj)) {
+            foreach ($obj as $role) {
+                $roleName = $role->getName();
+                if (is_string($needle)) {
+                    if ($roleName === $needle) {
+                        $rv++;
+                    }
+                } elseif (is_object($needle) && (get_class($needle) == 'Zend\Permissions\Rbac\Role')) {
+                    if ($roleName == $needle->getName()) {
+                        $rv++;
+                    }
+                }
+                $rv += $this->roleSearchIncludingChildren($role, $needle);
+            }
+        } else {
+            $children = $obj->getChildren();
+
+            // need to make sure the children are arrays (meaning they are added correctly)
+            if (! is_array($children)) {
+                return $rv ? true : false;
+            } elseif (! count($children)) {
+                if (is_object($needle) && (get_class($needle) == 'Zend\Permissions\Rbac\Role')) {
+                    if ($obj->getName() == $needle->getName()) {
+                        $rv++;
+                    }
+                }
+            } else {
+                foreach ($children as $child) {
+                    $roleName = $child->getName();
+                    if (is_string($needle)) {
+                        if ($roleName === $needle) {
+                            $rv++;
+                        }
+                    } elseif (is_object($needle) && (get_class($needle) == 'Zend\Permissions\Rbac\Role')) {
+                        if ($roleName == $needle->getName()) {
+                            $rv++;
+                        }
+                    }
+                    $rv += $this->roleSearchIncludingChildren($child, $needle);
+                }
+            }
```

Comment:
Just like the other block, we could do the following:

```
if (! is_array($children)) {
    return false;
}
if (! count($children)) {
    return is_object($needle) && $needle instanceof Role && $obje->getName() === $needle->getName();
}

foreach ($children as $child) {
    $roleName = $hcild->getName();
    if (is_string($needle) && $roleName === $needle) {
        return true;
    }
    if (is_object($needle) && $needle instanceof Role && $roleName === $needle->getName()) {
        return true;
    }
    if ($this->roleSearchIncludingChildren($child, $needle)) {
        return true;
    }
}
return false;
```

Making these changes and the previous ones I suggested would eliminate unnecessary cycles when we've already determined a matching child is present.